scipy/sparse/csgraph/_matching.pyx

```diff
@@ -273,3 +274,495 @@ cdef tuple _hopcroft_karp(ITYPE_t[:] indices, ITYPE_t[:] indptr,
+    kk : memoryview
+        Corresponds to the `data` attribute of the graph in CSR format.
+    nr : int
+        The number of rows of the graph in CSR format.
+    nc : int
+        The number of columns of the graph in CSR format. Must be at least as
+        large as `nr`.
+
+    Returns
+    -------
+    row_ind : memoryview of length :math:`|U|`
+        For each row, the column matched with that row.
+
+    References
+    ----------
+    .. [1] http://www.assignmentproblems.com/LAPJV.htm
+    .. [2] Roy Jonker and Anton Volgenant:
+       A Shortest Augmenting Path Algorithm for Dense and Sparse Linear
+       Assignment Problems.
+       Computing 38:325-340, 1987.
+
+    """
+    cdef ITYPE_t l0, jp, t, i, lp, j1, tp, j0p, j1p, l0p, h, i0, td1
+    cdef DTYPE_t min_diff, v0, vj, dj
+    cdef DTYPE_t[:] v = np.zeros(nc, dtype=DTYPE)
+    cdef ITYPE_t[:] x = np.empty(nr, dtype=ITYPE)
+    for i in range(nr):
+        x[i] = -1
+    cdef ITYPE_t[:] y = np.empty(nc, dtype=ITYPE)
+    for j in range(nc):
+        y[j] = -1
+    cdef DTYPE_t[:] u = np.zeros(nr, dtype=DTYPE)
+    cdef DTYPE_t[:] d = np.zeros(nc, dtype=DTYPE)
+    cdef BTYPE_t[:] ok = np.zeros(nc, dtype=BTYPE)
+    cdef BTYPE_t[:] xinv = np.zeros(nr, dtype=BTYPE)
+    cdef ITYPE_t[:] free = np.empty(nr, dtype=ITYPE)
+    for i in range(nr):
+        free[i] = -1
+    cdef ITYPE_t[:] todo = np.empty(nc, dtype=ITYPE)
+    for j in range(nc):
+        todo[j] = -1
+    cdef ITYPE_t[:] lab = np.zeros(nc, dtype=ITYPE)
+
+    # We skip the initialization entirely in the non-square case and instead
+    # fill all of `free` explicitly.
+    if nr == nc:
+        for jp in range(nc):
+            v[jp] = INFINITY
+        for i in range(nr):
+            for t in range(first[i], first[i + 1]):
+                jp = kk[t]
+                if cc[t] < v[jp]:
+                    v[jp] = cc[t]
+                    y[jp] = i
+        for jp in range(nc - 1, -1, -1):
+            i = y[jp]
+            if x[i] == -1:
+                x[i] = jp
+            else:
+                y[jp] = -1
+                # Here, the original Pascal code simply inverts the sign of x;
+                # as that doesn't play too well with zero-indexing, we
+                # explicitly keep track of uniqueness instead.
+                xinv[i] = 1
+        lp = 0
+        for i in range(nr):
+            if xinv[i] == 1:
+                continue
+            if x[i] != -1:
+                min_diff = INFINITY
+                j1 = x[i]
+                for t in range(first[i], first[i + 1]):
+                    jp = kk[t]
+                    if jp != j1:
+                        if cc[t] - v[jp] < min_diff:
+                            min_diff = cc[t] - v[jp]
+                u[i] = min_diff
+                tp = first[i]
+                while kk[tp] != j1:
+                    tp += 1
+                v[j1] = cc[tp] - min_diff
+            else:
+                free[lp] = i
+                lp += 1
```

Comment:
the two lines above. appear in opposite order in the pascal code. it works because the difference in indexing, but its a bit tricky that lp is not index-corrected and then saved by a lineswap. Perhaps a comment about the difference to the pascal code would be nice.

Reply:
Good catch. A comment is certainly needed.